### ISSUE TITLE
Enable data collection for articles, not article-event pairs

### ISSUE THREAD

**Issue by davidskalinder** [davidskalinder/mpeds-coder#4]
Several options here:

1. Proper way:  New database table at article-coder grain to store article description (and possibly future article-coder-level data.  Will need to add form outside of event-creator-block template and probably outside of event-creator template as well; will probably also need new JS handler and database query code (though it may not be too bad to adapt existing handlers for this).  New structure will need to be handled in pass 2 structure.
2. Database kludge:  Could add article-level data as a dummy event (such as field maximum, though no less in case we ever get there) to each article.  This would remove need for the new table and probably for the new handler, though would need to put existing description form template outside of the event-creater template/structure.  Will need to be handled in pass 2 structure.
3. Human kludge:  edit "Provide a short description of this article"  —> "Provide a short description of this article. Needed only once per article. If there are multiple events, you can write ditto or leave blank in subsequent codes"

Option 3 is doable in seconds and is currently what pass 2 expects, though will require indefinite pain from coders.  Options 1 and 2 both need changes to pass 2 apparatus so will need consultation from JL.

**Comment by matthewsmorgan** [comment]
In the most recent version of changes, I'm noticing that you MUST click "add event" in order to even type article-level description. This is a bit confusing in cases where there is not actual event, which sometimes happens. Can the article-level info be put in a separate place from the events? This way, if it's a case of an article with no event, the Pass 1 coder can describe the article and move on without coding? (I guess it would also be important to put the weed-out questions here, so we no that the "no event" reason is why nothing was coded)

**Comment by davidskalinder** [comment]
Yep, that's basically the plan.  I haven't changed anything about this one yet. :)

**Comment by alexhanna** [comment]
fwiw, this change should probably be implemented upstream as well.

**Comment by davidskalinder** [comment]
> fwiw, this change should probably be implemented upstream as well.

Okay, good to know.  I guess we haven't talked about how to decide which things should go upstream?  Maybe I should also try to be more careful about building on top of a clean version of the upstream repo whenever possible (rather than building on top of a bunch of BPP features that upstream doesn't want)...

**Comment by alexhanna** [comment]
I'm flagging it here so building can occur in an upstream branch.

**Comment by davidskalinder** [comment]
Okay thanks.  Do you want to monitor, er, all of our issues to see which ones upstream wants?  That sounds like a lot of work lol

**Comment by alexhanna** [comment]
I'm already creeping on all your issues so yes.

**Comment by davidskalinder** [comment]
Hah, if you say so!  Still though, I'll try to default to building on branches from pure upstream when possible.

**Comment by davidskalinder** [comment]
Okay, the plan for this:

Article-level forms should always be shown (unlike event-level forms that are only shown when added or edited); so form elements don't strictly need to be added by JS (as event forms are in ec.js).  However, for consistency's sake, it's probably best to use the same structure for articles as for events (and we might want the ability to add/remove the fields at some point.

But the event-creator template now actually loads the article text, so we might as well stay within that template.  We need to put in ec.js a new function like modifyEvent that ajaxes a controller function like modifyEvents (which lives at _load_event_block) and which populates a new template like event-creator-block and returns it as the ajax response.  The new js function then needs to add listeners for each item in the article form.  Unlike modifyEvent, it needs to be run at page load, not on clicking something.

Tl;dr, the things we need are:

1. New db definitions somewhere (database? models? both?) to make tables that store article info.  Table should be at coder-article-variable grain, so will need a new table (cf coder_event_creator)
1. A new template for the article form block (cf event-creator-block.html)
1. A new controller function to populate the template and return it as an ajax response (cf modifyEvents)
1. A new empty div in event-creator.html to carry the new stuff (cf div id="event-blocks")
1. A new JS function in ec.js to request ajax from the controller function and append it to the new carrier div (cf modifyEvent)
   1. New listeners within this function to listen for each type of article-level form element (cf modifyEvent listeners)
1. A new line in the main function at the bottom of ec.js to call the new JS function on page load instead of on click (cf // add event listener)

Do those items from the bottom up to allow testing at every stage (dump test text on the page -- working in the opposite order will be too hard to shoehorn into the page).

**Comment by davidskalinder** [comment]
Notes to self:

/- 6
/- 5
- 5.i
/- 4
- 3
- 2
- 1

Still to do:
- do 1 as written and run setup.py with everything but init_db commented out
- do item 3 by copying everything from old controller function to new and removing/changing as necessary
- do test version of 2 by doing same for ecb template
- do 5.i by doing same for js function
- do proper version of 2

**Comment by davidskalinder** [comment]
Status:

/- 6
/- 5
- 5.i
/- 4
- 3
- 2
/- 1

Still to do:
/- do 1 as written and run setup.py with everything but init_db commented out
- do item 3 by copying everything from old controller function to new and removing/changing as necessary
- do test version of 2 by doing same for ecb template
- do 5.i by doing same for js function
- do proper version of 2

**Comment by davidskalinder** [comment]
Mocked up template for item 2, but not sure how much of it will work long-term...  Controller function is still an ultrafragile mess, for unclear reasons.

**Comment by davidskalinder** [comment]
From meeting discussion:

Article-level checkboxes:
- No protests
- Protests but no black protests
- No events in the US
- No events after 1990

**Comment by davidskalinder** [comment]
Issue needs to be split between enabling article-level info and particular info to collect.

**Comment by davidskalinder** [comment]
Time for a revision:

1. DONE. New db definitions somewhere (database? models? both?) to make tables that store article info.  Table should be at coder-article-variable grain, so will need a new table (cf coder_event_creator)
1. IN PROGRESS. A new template for the article form block (cf event-creator-block.html)
1. DONE. A new controller function to populate the template and return it as an ajax response (cf modifyEvents)
1. DONE. A new empty div in event-creator.html to carry the new stuff (cf div id="event-blocks")
1. New handlers in shared.js to handle article-level form events
   1. New controller functions to do DB queries for the article-level form events
1. DONE. A new JS function in ec.js to request ajax from the controller function and append it to the new carrier div (cf modifyEvent)
   1. IN PROGRESS. New listeners within this function to listen for each type of article-level form element (cf modifyEvent listeners)
1. DONE. A new line in the main function at the bottom of ec.js to call the new JS function on page load instead of on click (cf // add event listener)

Stuff still to do can be limited to textareas and check-boxes.

**Comment by davidskalinder** [comment]
I _think_ this is now all done in branch enable_article_coding, with stuff to handle checkboxes, text inputs/areas, and probably radios (though this isn't tested) at the article level.  The article block template has working stuff, but some silly variables atm -- this needs to be cleaned up in a separate branch for #5, at which point both branches can be merged into testing.

**Comment by matthewsmorgan** [comment]
Not sure I fully understand this issue, not sure what I'm testing. If I understand this is separating out article and event coding in the UI, yes this is working.

**Comment by davidskalinder** [comment]
> Not sure I fully understand this issue, not sure what I'm testing. If I understand this is separating out article and event coding in the UI, yes this is working.

Yeah, sorry, lots of implementation-speak in this thread.  But I think your understanding is correct: the issue is about enabling us to collect, for example, article descriptions that are attached to the article itself, not to events within the article.  (So I think this resolves your [comment](https://github.com/davidskalinder/mpeds-coder/issues/4#issuecomment-562204542) above . . .)
